src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlRangeTranslator.cs

```diff
@@ -37,53 +40,140 @@ namespace Npgsql.EntityFrameworkCore.PostgreSQL.Query.ExpressionTranslators.Inte
 
-            case nameof(NpgsqlRangeExtensions.DoesNotExtendRightOf):
-                return new CustomBinaryExpression(expression.Arguments[0], expression.Arguments[1], "&<", typeof(bool));
+            case nameof(NpgsqlRangeExtensions.DoesNotExtendRightOf) when VersionAtLeast(9, 2):
+                return new CustomBinaryExpression(e.Arguments[0], e.Arguments[1], "&<", typeof(bool));
 
-            case nameof(NpgsqlRangeExtensions.DoesNotExtendLeftOf):
-                return new CustomBinaryExpression(expression.Arguments[0], expression.Arguments[1], "&>", typeof(bool));
+            case nameof(NpgsqlRangeExtensions.DoesNotExtendLeftOf) when VersionAtLeast(9, 2):
+                return new CustomBinaryExpression(e.Arguments[0], e.Arguments[1], "&>", typeof(bool));
 
-            case nameof(NpgsqlRangeExtensions.IsAdjacentTo):
-                return new CustomBinaryExpression(expression.Arguments[0], expression.Arguments[1], "-|-", typeof(bool));
+            case nameof(NpgsqlRangeExtensions.IsAdjacentTo) when VersionAtLeast(9, 2):
+                return new CustomBinaryExpression(e.Arguments[0], e.Arguments[1], "-|-", typeof(bool));
 
-            case nameof(NpgsqlRangeExtensions.Union):
-                return new CustomBinaryExpression(expression.Arguments[0], expression.Arguments[1], "+", expression.Arguments[0].Type);
+            case nameof(NpgsqlRangeExtensions.Union) when VersionAtLeast(9, 2):
+                return new CustomBinaryExpression(e.Arguments[0], e.Arguments[1], "+", e.Arguments[0].Type);
 
-            case nameof(NpgsqlRangeExtensions.Intersect):
-                return new CustomBinaryExpression(expression.Arguments[0], expression.Arguments[1], "*", expression.Arguments[0].Type);
+            case nameof(NpgsqlRangeExtensions.Intersect) when VersionAtLeast(9, 2):
+                return new CustomBinaryExpression(e.Arguments[0], e.Arguments[1], "*", e.Arguments[0].Type);
 
             case nameof(NpgsqlRangeExtensions.Except):
-                return new CustomBinaryExpression(expression.Arguments[0], expression.Arguments[1], "-", expression.Arguments[0].Type);
+                return new CustomBinaryExpression(e.Arguments[0], e.Arguments[1], "-", e.Arguments[0].Type);
+
+            case nameof(NpgsqlRangeExtensions.Merge) when VersionAtLeast(9, 5):
+                return new SqlFunctionExpression("range_merge", e.Type, new[] { e.Arguments[0], e.Arguments[1] });
+
+            default:
+                return null;
+            }
+        }
+
+        /// <inheritdoc />
+        [CanBeNull]
+        public Expression Translate(MemberExpression e)
+        {
+            var type = e.Member.DeclaringType;
+            if (type == null || !type.IsGenericType || type.GetGenericTypeDefinition() != typeof(NpgsqlRange<>))
+                return null;
+
+            switch (e.Member.Name)
+            {
+            case nameof(NpgsqlRange<int>.LowerBound) when VersionAtLeast(9, 2) && !e.Type.IsNullableType():
```

Comment:
I'd not have two cases for value/reference, but handle them in the same case. You can even create the `SqlFunctionExpression` once and conditionally wrap it based on whether the type is nullable.

Actually, thinking about this some more, is this really the desired behavior? In which cases does PostgreSQL return null for a bound? Do we really want them to zeros when it's an int range?

Reply:
That's a good point, I'll merge these cases.

More details below, but the short answer is that we can't return a database null as the bound for a range over CLR value types. So `Range<T>.LowerBound` needs a coalesce wrapper if `T` is non-nullable, because the range could be empty or the bound could be infinite.

I played around with a few ways to mitigate this, and this seemed to be the least hack-ish. That said, it's entirely possible that I've overlooked a simpler solution, so I'm open to any alternatives. (It would be nice not to need the coalesce wrapper...)

Reply:
I'll try to investigate a bit tomorrow.